test/test_init_teardown.c

```diff
@@ -0,0 +1,238 @@
+
+    ret = (*umfMemoryProviderCreateFromMemspace)(memspaceHostAll, NULL,
+                                                 &umf_provider_default);
+    if (ret || umf_provider_default == NULL) {
+        fprintf(stderr, "umf_load: error creating the default provider: %i\n",
+                ret);
+        goto err_dlclose;
+    }
+    fprintf(stderr, "umf_load: the default provider created from memspace\n");
+
+    umf_default = NULL;
+    ret = (*umfPoolCreate)((*umfScalablePoolOps)(), umf_provider_default, NULL,
+                           0, &umf_pool_default);
+    if (ret || umf_pool_default == NULL) {
+        fprintf(stderr, "umf_load: error creating the default pool: %i\n", ret);
+        goto err_destroy_provider;
+    }
+    fprintf(stderr,
+            "umf_load: the default pool created from the All Nodes provider\n");
+
+    umf_default = umf_pool_default; // umf pool using the default memspace
+
+    fprintf(stderr, "umf_load: umf initialized\n");
+
+    return 0;
+
+err_destroy_provider:
+    (*umfMemoryProviderDestroy)(umf_provider_default);
+err_dlclose:
+    dlclose(h_umf);
+
+    return -1;
+}
+
+static void umf_unload(void) {
+    umf_default = NULL;
+
+    fprintf(stderr, "umf_unload: finalizing UMF ...\n");
+
+    (*umfPoolDestroy)(umf_pool_default);
+    fprintf(stderr, "umf_unload: the default umf memory pool destroyed\n");
+
+    (*umfMemoryProviderDestroy)(umf_provider_default);
+    fprintf(stderr, "umf_unload: the default umf memory provider destroyed\n");
+
+    // deinitialize libumf (decrement the reference counter of users)
+    fprintf(stderr, "umf_unload: calling umfTearDown() ...\n");
+    (*umfTearDown)();
+
+    fprintf(stderr, "umf_unload: closing umf library ...\n");
+    dlclose(h_umf);
+    fprintf(stderr, "umf_unload: umf library closed\n");
+}
+
+static int run_test(int wrong_dtor_order) {
+
+    if (wrong_dtor_order) {
+        fprintf(stderr, "\n\n*** Running test with INCORRECT order of "
+                        "destructors ***\n\n\n");
+    } else {
+        fprintf(
+            stderr,
+            "\n\n*** Running test with CORRECT order of destructors ***\n\n\n");
+    }
+
+    if (umf_load()) {
+        return -1;
+    }
+
+    assert(umf_default);
+    void *ptr = (*umf_alloc)(umf_default, SIZE_ALLOC);
+    (*umf_free)(umf_default, ptr);
+
+    // simulate incorrect order of destructors (an additional, unwanted destructor call)
+    if (wrong_dtor_order) {
+        fprintf(stderr,
+                "*** Simulating incorrect order of destructors !!! ***\n");
+        (*umfTearDown)();
+    }
+
+    umf_unload();
+
+    return 0;
+}
+
+int main(void) {
+
+    // The libtbbmalloc.so.2 library is required to run this test,
+    // because it uses umfScalablePoolOps.
+    // Skip this test if the libtbbmalloc.so.2 library is not found.
+    void *tbb = dlopen("libtbbmalloc.so.2", RTLD_LAZY);
+    if (tbb == NULL) {
+        fprintf(stderr, "SKIP: required libtbbmalloc.so.2 library not found\n");
```

Comment:
couldn't you use CMake's var `UMF_POOL_SCALABLE_ENABLED` to skip this within CMake...?

Reply:
No, I cannot. `UMF_POOL_SCALABLE_ENABLED` can be `ON`, but TBB can be not installed in the OS.

Reply:
hmm.. I think we're doing exactly this with other tests

https://github.com/oneapi-src/unified-memory-framework/blob/main/CMakeLists.txt#L210
\+ enabling tests in CMake, only `if(UMF_POOL_SCALABLE_ENABLED)`

Reply:
Done.